Add layouts for split 3x5 and 3x6 keyboards

ddeklerk opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

This pull requests adds two layouts: split_3x5_3 and split_3x6_3. The naming goes as follows: "split_[rows]x[columns on one half]_[thumb keys on one half]". This naming convention also allows for layouts with more complex thumb clusters, for example "split_5x6_2x3". I'm not sure whether to change the 3 for the thumb keys to 1x3 in these examples.

The reason for creating these layouts is that I recently acquired a Centromere and Centromere Mini, whilst already owning a Corne and Minidox. The Centromere and Corne have the same amount of keys and layout, and the Centromere Mini and Minidox have the same amount of keys and layout.

I also intended on adding support for these layouts to the keyboards I mentioned, but I am unsure whether it's better to do that in a separate pull request.

In the other layouts I also saw JSON files describing the layouts, but I was not able to find anything about that in the documentation and don't know how to make those myself. I'd gladly include those as well once I know how to.

This is also my first contribution to QMK, so let me know if I can improve anything.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

Erovia commented:
> In the other layouts I also saw JSON files describing the layouts, but I was not able to find anything about that in the documentation and don't know how to make those myself. I'd gladly include those as well once I know how to.

There is a `layout.json` which is a [KLE](http://www.keyboard-layout-editor.com) output. I'm not sure if it has any use, but I guess could be useful to load the layout to said tool.

The other is `info.json`, which is used by many tools and is documented [here](https://docs.qmk.fm/#/reference_info_json).

ddeklerk commented:
A few questions regarding `info.json`.

With the layout being supported by keyboards with varying key sizes, would it be best to just stick with 1u keys for the most part? For example, the Centromere uses all 1u keys, but the Minidox and Crkbd have taller thumb keys and the Crkbd wider outer column keys.

How do I go about figuring out the width and height of the layout? I looked at the Ergodox layout as an example since it's the only other layout that's split or has tilted keys, but its dimensions did not match the ones I got in KLE when pasting its `layout.json`.

The documentation states that `layouts.key_count` is required, but I'm not seeing that included in the JSON files I'm using for reference or many of the other `info.json` files.

fauxpark commented:
You can generate an info.json by pasting the KLE raw JSON into https://qmk.fm/converter, and preview it in the Configurator by pressing Ctrl+Shift+I.

fauxpark commented on layouts/default/split_3x5_3/default_split_3x5_3/keymap.c:
```diff
@@ -0,0 +1,14 @@
+#include QMK_KEYBOARD_H
+
+enum layers {
+    _QWERTY
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [_QWERTY] = LAYOUT_split_3x5_3(
+        KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,                         KC_Y,    KC_U,    KC_I,    KC_O,   KC_P,\
+        KC_A,    KC_S,    KC_D,    KC_F,    KC_G,                         KC_H,    KC_J,    KC_K,    KC_L, KC_SCLN,\
+        KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,                         KC_N,    KC_M, KC_COMM,  KC_DOT, KC_SLSH,\
+                                        KC_LGUI,   KC_BSPC,  KC_SPC,     KC_SPC,  KC_ENT,  KC_RALT \
+    )
+};
```
```suggestion
const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
    [0] = LAYOUT_split_3x5_3(
        KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,                         KC_Y,    KC_U,    KC_I,    KC_O,   KC_P,
        KC_A,    KC_S,    KC_D,    KC_F,    KC_G,                         KC_H,    KC_J,    KC_K,    KC_L, KC_SCLN,
        KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,                         KC_N,    KC_M, KC_COMM,  KC_DOT, KC_SLSH,
                                        KC_LGUI,   KC_BSPC,  KC_SPC,     KC_SPC,  KC_ENT,  KC_RALT
    )
};
```

ddeklerk replied:
I can see that the layers enum has been removed and the layer is simply 0, but what's different about the layout itself? Looks identical to me.

fauxpark replied:
I've removed the backslashes from the end of the lines, they're not needed.

ddeklerk replied:
I see. I'll remember that for the future. I think a bunch of keymaps still have those, since I copied the basis for this from Centromere.

fauxpark commented on layouts/default/split_3x6_3/default_split_3x6_3/keymap.c:
```diff
@@ -0,0 +1,14 @@
+#include QMK_KEYBOARD_H
+
+enum layers {
+    _QWERTY
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [_QWERTY] = LAYOUT_split_3x6_3(
+        KC_TAB,     KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,                         KC_Y,    KC_U,    KC_I,    KC_O,   KC_P,  KC_BSPC,\
+        KC_LCTL,    KC_A,    KC_S,    KC_D,    KC_F,    KC_G,                         KC_H,    KC_J,    KC_K,    KC_L, KC_SCLN, KC_QUOT,\
+        KC_LSFT,    KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,                         KC_N,    KC_M, KC_COMM,  KC_DOT, KC_SLSH, KC_RSFT,\
+                                        KC_LGUI,   KC_BSPC,  KC_SPC,     KC_SPC,  KC_ENT,  KC_RALT \
+    )
+};
```
```suggestion
const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
    [0] = LAYOUT_split_3x6_3(
        KC_TAB,     KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,                         KC_Y,    KC_U,    KC_I,    KC_O,   KC_P,  KC_BSPC,
        KC_LCTL,    KC_A,    KC_S,    KC_D,    KC_F,    KC_G,                         KC_H,    KC_J,    KC_K,    KC_L, KC_SCLN, KC_QUOT,
        KC_LSFT,    KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,                         KC_N,    KC_M, KC_COMM,  KC_DOT, KC_SLSH, KC_RSFT,
                                        KC_LGUI,   KC_BSPC,  KC_SPC,     KC_SPC,  KC_ENT,  KC_RALT
    )
};
```

ddeklerk commented:
I have now added and tested the JSON files as best I could.

I did run into a minor bug with the converter. Some numbers were changed with many trailing zeroes and a one. So `1.38` became `1.3800000000000001`. I fixed those before committing of course.

fauxpark commented on layouts/default/split_3x5_3/layout.json:
```diff
@@ -0,0 +1,21 @@
+[{x:2,a:7},"",{x:7},""],
+[{y:-0.875,x:1},"",{x:1},""],
+[{y:-0.995,x:9},"",{x:1},""],
+[{y:-0.88,x:4},""],
+[{y:-0.875},""],
+[{y:-0.995,x:8},"",{x:3},""],
+[{y:-0.38,x:2},"",{x:7},""],
+[{y:-0.875,x:1},"",{x:1},""],
+[{y:-0.995,x:9.01},"",{x:0.99},""],
+[{y:-0.88,x:4},""],
+[{y:-0.875},""],
+[{y:-0.995,x:8},"",{x:3},""],
+[{y:-0.38,x:2},"",{x:7},""],
+[{y:-0.875,x:1},"",{x:1},""],
+[{y:-0.995,x:9},"",{x:1},""],
+[{y:-0.88,x:4},""],
+[{y:-0.875},""],
+[{y:-0.995,x:8},"",{x:3},""],
+[{y:-0.13,x:2.5},"",{x:6},""],
+[{y:-0.75,x:3.5},"",{x:4},""],
+[{y:-0.75,x:4.5},"",{x:2},""]
```
```suggestion
[{x:2,a:7},"",{x:6},""],
[{y:-0.875,x:1},"",{x:1},"",{x:4},"",{x:1},""],
[{y:-0.875},"",{x:3},"",{x:2},"",{x:3},""],
[{y:-0.25,x:2},"",{x:6},""],
[{y:-0.875,x:1},"",{x:1},"",{x:4},"",{x:1},""],
[{y:-0.875},"",{x:3},"",{x:2},"",{x:3},""],
[{y:-0.25,x:2},"",{x:6},""],
[{y:-0.875,x:1},"",{x:1},"",{x:4},"",{x:1},""],
[{y:-0.875},"",{x:3},"",{x:2},"",{x:3},""],
[{x:2.5},"",{x:5},""],
[{y:-0.75,x:3.5},"",{x:3},""],
[{y:-0.75,x:4.5},"",{x:1},""]
```

fauxpark commented on layouts/default/split_3x6_3/layout.json:
```diff
@@ -0,0 +1,15 @@
+[{x:3,a:7},"",{x:7},""],
+[{y:-0.87,x:2},"",{x:1},"",{x:5},"",{x:1},""],
+[{y:-0.88,x:5},""],
+[{y:-0.87},"","",{x:7},"",{x:3},"",""],
+[{y:-0.38,x:3},"",{x:7},""],
+[{y:-0.87,x:2},"",{x:1},"",{x:5.01},"",{x:0.99},""],
+[{y:-0.88,x:5},""],
+[{y:-0.87},"","",{x:7},"",{x:3},"",""],
+[{y:-0.38,x:3},"",{x:7},""],
+[{y:-0.87,x:2},"",{x:1},"",{x:5},"",{x:1},""],
+[{y:-0.88,x:5},""],
+[{y:-0.87},"","",{x:7},"",{x:3},"",""],
+[{y:-0.13,x:3.5},"",{x:6},""],
+[{y:-0.75,x:4.5},"",{x:4},""],
+[{y:-0.75,x:5.5},"",{x:2},""]
```
```suggestion
[{x:3,a:7},"",{x:6},""],
[{y:-0.875,x:2},"",{x:1},"",{x:4},"",{x:1},""],
[{y:-0.875},"","",{x:3},"",{x:2},"",{x:3},"",""],
[{y:-0.25,x:3},"",{x:6},""],
[{y:-0.875,x:2},"",{x:1},"",{x:4},"",{x:1},""],
[{y:-0.875},"","",{x:3},"",{x:2},"",{x:3},"",""],
[{y:-0.25,x:3},"",{x:6},""],
[{y:-0.875,x:2},"",{x:1},"",{x:4},"",{x:1},""],
[{y:-0.875},"","",{x:3},"",{x:2},"",{x:3},"",""],
[{x:3.5},"",{x:5},""],
[{y:-0.75,x:4.5},"",{x:3},""],
[{y:-0.75,x:5.5},"",{x:1},""]
```

fauxpark commented on layouts/default/split_3x5_3/info.json:
```diff
@@ -0,0 +1,49 @@
+    "layouts": {
+        "LAYOUT_split_3x5_3": {
+            "layout": [
+                {"x":2, "y":0},
+                {"x":10, "y":0},
+                {"x":1, "y":0.125},
+                {"x":3, "y":0.125},
+                {"x":9, "y":0.13},
+                {"x":11, "y":0.13},
+                {"x":4, "y":0.25},
+                {"x":0, "y":0.375},
+                {"x":8, "y":0.38},
+                {"x":12, "y":0.38},
+                {"x":2, "y":1.0},
+                {"x":10, "y":1.0},
+                {"x":1, "y":1.125},
+                {"x":3, "y":1.125},
+                {"x":9.01, "y":1.13},
+                {"x":11.0, "y":1.13},
+                {"x":4, "y":1.25},
+                {"x":0, "y":1.375},
+                {"x":8, "y":1.38},
+                {"x":12, "y":1.38},
+                {"x":2, "y":2},
+                {"x":10, "y":2},
+                {"x":1, "y":2.125},
+                {"x":3, "y":2.125},
+                {"x":9, "y":2.13},
+                {"x":11, "y":2.13},
+                {"x":4, "y":2.25},
+                {"x":0, "y":2.375},
+                {"x":8, "y":2.38},
+                {"x":12, "y":2.38},
+                {"x":2.5, "y":3.25},
+                {"x":9.5, "y":3.25},
+                {"x":3.5, "y":3.5},
+                {"x":8.5, "y":3.5},
+                {"x":4.5, "y":3.75},
+                {"x":7.5, "y":3.75}
```
The ordering is incorrect here. KLE orders by X/Y coordinates, whereas info.json goes by the actual ordering of the elements in the array. So sometimes you may need to reorder them after conversion - you can test this by turning on fast input in the Configurator and spamming a key.

I've also tweaked the KLE to be properly symmetrical and aligned on 0.125u intervals, which avoids the rounding errors in the converter.
```suggestion
                {"x":0, "y":0.25},
                {"x":1, "y":0.125},
                {"x":2, "y":0},
                {"x":3, "y":0.125},
                {"x":4, "y":0.25},

                {"x":7, "y":0.25},
                {"x":8, "y":0.125},
                {"x":9, "y":0},
                {"x":10, "y":0.125},
                {"x":11, "y":0.25},

                {"x":0, "y":1.25},
                {"x":1, "y":1.125},
                {"x":2, "y":1},
                {"x":3, "y":1.125},
                {"x":4, "y":1.25},

                {"x":7, "y":1.25},
                {"x":8, "y":1.125},
                {"x":9, "y":1},
                {"x":10, "y":1.125},
                {"x":11, "y":1.25},

                {"x":0, "y":2.25},
                {"x":1, "y":2.125},
                {"x":2, "y":2},
                {"x":3, "y":2.125},
                {"x":4, "y":2.25},

                {"x":7, "y":2.25},
                {"x":8, "y":2.125},
                {"x":9, "y":2},
                {"x":10, "y":2.125},
                {"x":11, "y":2.25},

                {"x":2.5, "y":3.25},
                {"x":3.5, "y":3.5},
                {"x":4.5, "y":3.75},

                {"x":6.5, "y":3.75},
                {"x":7.5, "y":3.5},
                {"x":8.5, "y":3.25}
```

fauxpark commented on layouts/default/split_3x6_3/info.json:
```diff
@@ -0,0 +1,55 @@
+                {"x":2, "y":0.13},
+                {"x":4, "y":0.13},
+                {"x":10, "y":0.13},
+                {"x":12, "y":0.13},
+                {"x":5, "y":0.25},
+                {"x":0, "y":0.38},
+                {"x":1, "y":0.38},
+                {"x":9, "y":0.38},
+                {"x":13, "y":0.38},
+                {"x":14, "y":0.38},
+                {"x":3, "y":1.0},
+                {"x":11, "y":1.0},
+                {"x":2, "y":1.13},
+                {"x":4, "y":1.13},
+                {"x":10.01, "y":1.13},
+                {"x":12.0, "y":1.13},
+                {"x":5, "y":1.25},
+                {"x":0, "y":1.38},
+                {"x":1, "y":1.38},
+                {"x":9, "y":1.38},
+                {"x":13, "y":1.38},
+                {"x":14, "y":1.38},
+                {"x":3, "y":2},
+                {"x":11, "y":2},
+                {"x":2, "y":2.13},
+                {"x":4, "y":2.13},
+                {"x":10, "y":2.13},
+                {"x":12, "y":2.13},
+                {"x":5, "y":2.25},
+                {"x":0, "y":2.38},
+                {"x":1, "y":2.38},
+                {"x":9, "y":2.38},
+                {"x":13, "y":2.38},
+                {"x":14, "y":2.38},
+                {"x":3.5, "y":3.25},
+                {"x":10.5, "y":3.25},
+                {"x":4.5, "y":3.5},
+                {"x":9.5, "y":3.5},
+                {"x":5.5, "y":3.75},
+                {"x":8.5, "y":3.75}
```
```suggestion
                {"x":0, "y":0.25},
                {"x":1, "y":0.25},
                {"x":2, "y":0.125},
                {"x":3, "y":0},
                {"x":4, "y":0.125},
                {"x":5, "y":0.25},

                {"x":8, "y":0.25},
                {"x":9, "y":0.125},
                {"x":10, "y":0},
                {"x":11, "y":0.125},
                {"x":12, "y":0.25},
                {"x":13, "y":0.25},

                {"x":0, "y":1.25},
                {"x":1, "y":1.25},
                {"x":2, "y":1.125},
                {"x":3, "y":1},
                {"x":4, "y":1.125},
                {"x":5, "y":1.25},

                {"x":8, "y":1.25},
                {"x":9, "y":1.125},
                {"x":10, "y":1},
                {"x":11, "y":1.125},
                {"x":12, "y":1.25},
                {"x":13, "y":1.25},

                {"x":0, "y":2.25},
                {"x":1, "y":2.25},
                {"x":2, "y":2.125},
                {"x":3, "y":2},
                {"x":4, "y":2.125},
                {"x":5, "y":2.25},

                {"x":8, "y":2.25},
                {"x":9, "y":2.125},
                {"x":10, "y":2},
                {"x":11, "y":2.125},
                {"x":12, "y":2.25},
                {"x":13, "y":2.25},

                {"x":3.5, "y":3.25},
                {"x":4.5, "y":3.5},
                {"x":5.5, "y":3.75},

                {"x":7.5, "y":3.75},
                {"x":8.5, "y":3.5},
                {"x":9.5, "y":3.25}
```

ddeklerk commented:
Keen eye, fauxpark! There's no way to commit the suggested changes without committing them one by one is there?

Oh, and I still would like some input regarding the addition of support for these layouts to the keyboards I mentioned (and possibly more). Should I tackle that in a separate pull request after this one has been merged?

fauxpark commented:
If you switch to the files tab, you can batch suggestions.

As for the layout support, yes, that should be done in a separate PR.

ddeklerk commented:
> As for the layout support, yes, that should be done in a separate PR.

One pull request for all the keyboards I want to add support for, or a pull request per keyboard?

fauxpark commented on layouts/default/split_3x5_3/info.json:
```diff
@@ -0,0 +1,56 @@
+    "keyboard_name": "3x5+3 split layout",
+    "url": "",
+    "maintainer": "qmk",
+    "width": 13,
```
```suggestion
    "width": 12,
```

fauxpark commented on layouts/default/split_3x6_3/info.json:
```diff
@@ -0,0 +1,62 @@
+    "keyboard_name": "3x6+3 split layout",
+    "url": "",
+    "maintainer": "qmk",
+    "width": 15,
```
```suggestion
    "width": 14,
```

fauxpark commented:
> One pull request for all the keyboards I want to add support for, or a pull request per keyboard?

Depends how large the changeset is.

Erovia commented on layouts/default/split_3x5_3/readme.md:
```diff
@@ -0,0 +1,3 @@
+# split_3x5_3
+
+   LAYOUT_split_3x5_3
```
```suggestion
   LAYOUT_split_3x5_3
   
   Layout macro for 3x5 split keyboards with 3 keys thumb cluster.
```
I think a description would should be provided to help others understand the naming. You can probably come up with something better than me.

Erovia commented on layouts/default/split_3x6_3/readme.md:
```diff
@@ -0,0 +1,3 @@
+# split_3x6_3
+
+   LAYOUT_split_3x6_3
```
```suggestion
   LAYOUT_split_3x6_3
   
   Layout macro for 3x6 split keyboards with 3 keys thumb cluster.
```
I think a description would should be provided to help others understand the naming. You can probably come up with something better than me.

fauxpark commented on layouts/default/split_3x5_3/default_split_3x5_3/keymap.c:
```diff
@@ -0,0 +1,10 @@
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT_split_3x5_3(
```
Almost forgot...
```suggestion
     /*
      * ┌───┬───┬───┬───┬───┐       ┌───┬───┬───┬───┬───┐
      * │ Q │ W │ E │ R │ T │       │ Y │ U │ I │ O │ P │
      * ├───┼───┼───┼───┼───┤       ├───┼───┼───┼───┼───┤
      * │ A │ S │ D │ F │ G │       │ H │ J │ K │ L │ ; │
      * ├───┼───┼───┼───┼───┤       ├───┼───┼───┼───┼───┤
      * │ Z │ X │ C │ V │ B │       │ N │ M │ , │ . │ / │
      * └───┴───┴───┴───┴───┘       └───┴───┴───┴───┴───┘
      *           ┌───┐                   ┌───┐
      *           │GUI├───┐           ┌───┤Alt│
      *           └───┤Bsp├───┐   ┌───┤Ent├───┘
      *               └───┤   │   │   ├───┘
      *                   └───┘   └───┘
      */
    [0] = LAYOUT_split_3x5_3(
```

fauxpark commented on layouts/default/split_3x6_3/default_split_3x6_3/keymap.c:
```diff
@@ -0,0 +1,10 @@
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT_split_3x6_3(
```
```suggestion
     /*
      * ┌───┬───┬───┬───┬───┬───┐       ┌───┬───┬───┬───┬───┬───┐
      * │Tab│ Q │ W │ E │ R │ T │       │ Y │ U │ I │ O │ P │Bsp│
      * ├───┼───┼───┼───┼───┼───┤       ├───┼───┼───┼───┼───┼───┤
      * │Ctl│ A │ S │ D │ F │ G │       │ H │ J │ K │ L │ ; │ ' │
      * ├───┼───┼───┼───┼───┼───┤       ├───┼───┼───┼───┼───┼───┤
      * │Sft│ Z │ X │ C │ V │ B │       │ N │ M │ , │ . │ / │Sft│
      * └───┴───┴───┴───┴───┴───┘       └───┴───┴───┴───┴───┴───┘
      *               ┌───┐                   ┌───┐
      *               │GUI├───┐           ┌───┤Alt│
      *               └───┤Bsp├───┐   ┌───┤Ent├───┘
      *                   └───┤   │   │   ├───┘
      *                       └───┘   └───┘
      */
    [0] = LAYOUT_split_3x6_3(
```

manna-harbour commented:
Thanks for creating this PR.

Other than the already mentioned minidox and centromere mini, other 5x3_3 keyboards in QMK are [crkbd](https://github.com/qmk/qmk_firmware/tree/master/keyboards/crkbd) (corne mini version), [miniaxe](https://github.com/qmk/qmk_firmware/tree/master/keyboards/miniaxe), [gergoplex](https://github.com/qmk/qmk_firmware/pull/8928), and [squiggle](https://github.com/qmk/qmk_firmware/pull/9387) (minidox version).  I'm happy to help with adding these.  The maintainers should be asked first though.

ddeklerk commented:
Regarding the readme files, I am happy to write a little more about what the layouts are for and such, but at the same time almost all of the layouts already in the repository have a simple readme like the ones I have included already.

I think for the support for keyboards of this layout it would be best to at least split them up in two pull requests between the two layouts. I don't know whether the maintainers should be asked first, though. Is that the way it normally goes? I think I read in the docs that the maintainer should be tagged in the pull request, but nothing besides that.

Erovia commented:
> almost all of the layouts already in the repository have a simple readme

Their names are way easier to "decode", though.

> at least split them up in two pull requests between the two layouts

Agreed.
